### ISSUE TITLE
Open references if definition result has the same range than clicked offset

### ISSUE THREAD

**Issue by InSyncWithFoo** [redhat-developer/lsp4ij#531]
Currently <kbd>Ctrl</kbd> + Click always send `textDocument/definition`, which might not be ideal when the definition itself is clicked. LSP4IJ integrators should be able to configure that programmatically.

(Filed as requested at [this issue](https://github.com/InSyncWithFoo/pyright-for-pycharm/issues/72).)

**Comment by CppCXY** [comment]
in VSCode, when you ctrl + click on a reference, it first sends a textDocument/definition request. If the returned range overlaps with the clicked position, it then sends a textDocument/references request.

**Comment by angelozerr** [comment]
Thanks so much @CppCXY for your information.

I see now more the benefit to consume references.

@InSyncWithFoo is it the same usecase in the reported issue from your project?

**Comment by angelozerr** [comment]
> Yes, that's probably the best behaviour.

Indeed, but do you think it is th ebehavior than user would like to have?

**Comment by angelozerr** [comment]
@InSyncWithFoo after debugging Java support in IJ, I undertand more how the go to declaration is working.

There are 2 means to manage go to declaration:

 * extends GotoDeclarationHandler like we did with LSPGotoDeclarationHandler. The advantage to use that is if there are some custom PsiFile (it is your case @InSyncWithFoo I think), the go to declaration will take the priority and if another plugin which manages go to declaration like PyCharm I think,the LSP go to definition will work. The inconevinant to use this GotoDeclarationHandler are:
   * the hyperlinked ignore the LSP location link and with textmate you have this issue  https://github.com/redhat-developer/lsp4ij/issues/98
   * there is no way to manage the click action, so it is impossible to Show the References panel.
* using the PsiFile to collect declarations / references. In the LSP context, we have no custom PsiFile BUT it is possible to get declarations / referernces by implementing PsiSymbolDeclarationProvider / ImplicitReferenceProvider. But there are 2 problems:
  *  ImplicitReferenceProvider#getImplicitReferences is called every time (when you do an hoverfor instance), so if we implement textDocument/definition in the getImplicitReferences, it will consume textDocument/definitionevery time. In our case we need to execute getImplicitReferences  only when a GoToAction is executed. So we could implement an action listener before/after to enable/disable the call of getImplicitReferences 
  * ImplicitReferenceProvider#getImplicitReferences is called only if PsiFile doesn't return references from teh given offset, it means in your case @InSyncWithFoo that you will loose teh LSP go to definition if the hyperlinked PsiElement is done.

What do you think about that @InSyncWithFoo : loosing go to definition (but I think PyCharm should work good) but benefit with Show references panel (and for other language the proper hyperlinked). For many language I think using ImplicitReferenceProvider#getImplicitReferences will be very nice,but before doing that I would liketo know your feeling. If you don't want to loose the LSP go to definition, perhaps we should provide an API to know if the language server must use GotoDeclarationHandler  or ImplicitReferenceProvider.

**Comment by SCWells72** [comment]
PR for this is now open as #853.

**Comment by angelozerr** [comment]
@InSyncWithFoo I think this issue is fixed, right?

**Comment by angelozerr** [comment]
Ok that's bad, I though @SCWells72 contribution fixed that.

**Comment by SCWells72** [comment]
> Ok that's bad, I though [@SCWells72](https://github.com/SCWells72) contribution fixed that.

In order for **Navigate | Declaration or Usages** to work properly when invoked on a declaration, the language server must provide semantic tokens for declaration name identifiers that have semantic token modifiers of `declaration` or `definition`. Is that the case here? Can you share the output of the semantic token inspector for a sample file that reproduces this with both semantic token types and modifiers displayed?

**Comment by angelozerr** [comment]
@noklam if I remember the problem is when your file have not a proper PsiFile (like TEXT, texmate) it doesn't work perfectly. @SCWells72 fixes that by implementing a custom view provider based on semantic tokens for TEX and textmate language.

Is it your usecase?

**Comment by angelozerr** [comment]
> How would implement a custom view provider helps? (sorry I am not too familiar with these concepts yet).

It is not a trivial issue and to be honnest with you I don't masterize this topic. Th ehard work has been done by @SCWells72 who uses semantic tokens information to provide an IntelliJ view provider which supports a lot of usecases like cmd+click which call references.

I don't remember why IntelliJ doesn't trigger reference without semantic tokens. I need to find time to understand again the real problem.
